quantus/metrics/axiomatic_metrics.py

```diff
@@ -521,133 +563,123 @@ def __init__(self, *args, **kwargs):
+        >> scores = metric(model=model, x_batch=x_batch, y_batch=y_batch, a_batch=a_batch_saliency, **{})
         """
-        # Reshape input batch to channel first order:
-        if "channel_first" in kwargs and isinstance(kwargs["channel_first"], bool):
-            channel_first = kwargs.get("channel_first")
-        else:
-            channel_first = utils.infer_channel_first(x_batch)
-        x_batch_s = utils.make_channel_first(x_batch, channel_first)
-
-        # Wrap the model into an interface.
-        if model:
-            model = utils.get_wrapped_model(model, channel_first)
-
-        # Update kwargs.
-        self.kwargs = {
+        return super().__call__(
+            model=model,
+            x_batch=x_batch,
+            y_batch=y_batch,
+            a_batch=a_batch,
+            s_batch=s_batch,
+            channel_first=channel_first,
+            explain_func=explain_func,
+            explain_func_kwargs=explain_func_kwargs,
+            device=device,
+            model_predict_kwargs=model_predict_kwargs,
             **kwargs,
-            **{k: v for k, v in self.__dict__.items() if k not in ["args", "kwargs"]},
-        }
-
-        # Run deprecation warnings.
-        warn_func.deprecation_warnings(self.kwargs)
-
-        self.last_results = []
-
-        explain_func = self.kwargs.get("explain_func", Callable)
-        asserts.assert_explain_func(explain_func=explain_func)
-        if a_batch is None:
-
-            # Generate explanations.
-            a_batch = explain_func(
-                model=model.get_model(),
-                inputs=x_batch,
-                targets=y_batch,
-                **self.kwargs,
-            )
+        )
 
-        # Expand attributions to input dimensionality.
-        a_batch = utils.expand_attribution_channel(a_batch, x_batch_s)
+    def process_instance(
+            self,
+            model: ModelInterface,
+            x: np.ndarray,
+            y: np.ndarray,
+            a: np.ndarray,
+            s: np.ndarray,
+    ):
+
+        x_shifted = self.perturb_func(
+            arr=x,
+            indices=np.arange(0, x.size),
+            indexed_axes=np.arange(0, x.ndim),
+            **self.perturb_func_kwargs,
+        )
+        x_shifted = model.shape_input(x_shifted, x.shape, channel_first=True)
+        asserts.assert_perturbation_caused_change(x=x, x_perturbed=x_shifted)
 
-        # Asserts.
-        asserts.assert_attributions(a_batch=a_batch, x_batch=x_batch)
+        # Generate explanation based on shifted input x.
+        a_shifted = self.explain_func(
+            model=model.get_model(), inputs=x_shifted, targets=y,
+            **self.explain_func_kwargs
+        )
 
-        # Use tqdm progressbar if not disabled.
-        if not self.display_progressbar:
-            iterator = zip(x_batch_s, y_batch, a_batch)
+        # Check if explanation of shifted input is similar to original.
+        if (a.flatten() != a_shifted.flatten()).all():
```

Comment:
Sidenote unrelated to this PR: This seems to be a really strict condition. Does it make sense to give some options here w.r.t. strictness (e.g., similarity_func(a, a_shifted, precision=something)). Also for similarly strict metrics.